rust/chains/abacus-ethereum/src/trait_builder.rs

```diff
@@ -0,0 +1,59 @@
+use std::sync::Arc;
+
+use ethers::prelude::*;
+
+use abacus_core::{ContractLocator, Signers};
+
+use crate::Connection;
+
+async fn make_signing_provider<M: Middleware>(
+    provider: M,
+    signer: Signers,
+) -> Result<SignerMiddleware<NonceManagerMiddleware<M>, Signers>, M::Error> {
+    let provider_chain_id = provider.get_chainid().await?;
+    let signer = ethers::signers::Signer::with_chain_id(signer, provider_chain_id.as_u64());
+
+    let address = ethers::prelude::Signer::address(&signer);
+    let provider = NonceManagerMiddleware::new(provider, address);
+
+    let signing_provider = SignerMiddleware::new(provider, signer);
+    Ok(signing_provider)
+}
+
+pub(crate) trait MakeableWithProvider {
+    type Output;
+
+    fn make<M: Middleware + 'static>(self, provider: M, locator: &ContractLocator) -> Self::Output;
+}
+
+pub(super) async fn build_trait<M: MakeableWithProvider>(
+    conn: Connection,
+    locator: &ContractLocator,
+    signer: Option<Signers>,
+    builder: M,
+) -> eyre::Result<M::Output> {
+    Ok(match conn {
+        Connection::Http { url } => {
+            let provider: crate::RetryingProvider<Http> = url.parse()?;
+            let provider = Arc::new(Provider::new(provider));
+
+            if let Some(signer) = signer {
```

Comment:
Not a big deal but noticing lines 40-46 here are the same as 51-57, could be an opportunity to pull it out of the match

Reply:
I thought that as well, but if you look very very carefully, all of the lines are unique and it is impossible to dedup without using a macro. The reason is each `builder.make` is getting called with different generic arguments. They are just implicit so it is hard to see without the IDE.

Reply:
Also FTR I am okay with code duplication if it saves on complexity. My biggest personal requirement is if you are going to duplicate code, keep the duplicates together.